index.js

```diff
@@ -164,121 +166,101 @@ FeatureService.prototype.getObjectIdField = function (info) {
-      }
-
-      // build where clause based pages
-      if (serviceInfo.supportsStatistics) {
-        this.statistics(serviceInfo.objectIdField, ['min', 'max'], function (err, stats) {
-          if (err) {
-            return callback(err)
-          }
-
-          try {
-            if (stats.error) {
-              try {
-                var idUrl = this.url + '/' + (this.layer || 0) + '/query?where=1=1&returnIdsOnly=true&f=json'
-                this.request(idUrl, function (err, idJson) {
-                  if (err) {
-                    return callback(err)
-                  }
-
-                  var minID, maxID
-                  if (idJson.error) {
-                    // DMF: if grabbing objectIDs fails fall back to guessing based on 0 and count
-                    minID = 0
-                    maxID = count
-                  } else {
-                    idJson.objectIds.sort(function (a, b) { return a - b })
-                    minID = idJson.objectIds[0]
-                    maxID = idJson.objectIds[idJson.objectIds.length - 1]
-                  }
-                  return callback(null, this._objectIdPages(minID, maxID, maxCount))
-                }.bind(this))
-
-              } catch (e) {
-                return callback(e)
-              }
-            } else {
-              var names, minId, maxId
-              var attrs = stats.features[0].attributes
-              if (stats && stats.fieldAliases) {
-                names = Object.keys(stats.fieldAliases)
-              }
-              minId = attrs.min_oid || attrs.MIN_OID || attrs[names[0]]
-              maxId = attrs.max_oid || attrs.MAX_OID || attrs[names[1]]
-
-              return callback(null, this._objectIdPages(minId, maxId, maxCount))
-            }
-          } catch (e) {
-            return callback(e)
-          }
+  this.metadata(function (err, meta) {
+    if (err) return callback(err)
+    var size = meta.size
+    var layer = meta.layer
+    this.options.objectIdField = meta.oid
+    size = Math.min(parseInt(size, 10), 1000) || 1000
+    var nPages = Math.ceil(meta.count / size)
+
+    // if the service supports paging, we can use offset to build pages
+    var canPage = layer.advancedQueryCapabilities && layer.advancedQueryCapabilities.supportsPagination
+    if (canPage) return callback(null, this._offsetPages(nPages, size))
+
+    // if the service supports statistics, we can request the maximum and minimum id to build pages
+    if (layer.supportsStatistics) {
+      this._getIdRangeFromStats(meta, function (err, stats) {
+      // if this worked then we can pagination using where clauses
+        if (!err) return callback(null, this._rangePages(stats, size))
+        // if it failed, try to request all the ids and split them into pages
+        this.layerIds(function (err, ids) {
+          // either this works or we give up
+          return callback(err, this._idPages(ids, size))
```

Comment:
same as above. `err` implies `ids` should be null, so `this._idPages` shouldn't be called.